// gupshup-local-botrunner/lib/BotUtils.js
exports.isJson = isJson = function (jsonStr) {
    try {
        if (typeof jsonStr == 'object') {
            return true;
        }
        JSON.parse(jsonStr);
    } catch (e) {
        return false;
    }
    return true;
}

exports.parseJson = parseJson = function (v) {
    try {
        return JSON.parse(v);
    } catch (e) {
        return v;
    }
}

exports.isEmpty = isEmpty = function (obj) {
    if (typeof obj === "undefined")
        return true;
    return Object.keys(obj).length === 0;
}

exports.errorResponse = errorResponse = {
    code: 404,
    message: "The requested resource was not found"
};

exports.getStringForm = getStringForm = function (obj) {
    if (typeof obj === "undefined") {
        return null;
    } else if (typeof obj === "number") {
        return obj + "";
    } else if (typeof obj === "boolean") {
        return obj + "";
    } else if (typeof obj === "string") {
        return obj;
    } else if (typeof obj === "object") {
        return JSON.stringify(obj);
    } else {
        return null;
    }
}

exports.setupEvent = function (event) {
    let botEvent = {};
    if (event.query) {
        botEvent.botname = event.params['botname'];
        Object.keys(event.query).map(function (key) {
            let val = event.query[key];
            val = parseJson(val);
            botEvent[key] = val;
        });
        if (botEvent.contextobj && botEvent.contextobj.contextid) {
            botEvent.sender = botEvent.contextobj.contextid;
        }
        if (botEvent.messageobj && botEvent.messageobj.text) {
            botEvent.message = botEvent.messageobj.text;
        }
    } else {
        Object.keys(event).map(function (key) {
            let val = event[key];
            val = (val);

            botEvent[key] = val;
        });
        if (event.query && event.query.message) {
            botEvent.message = event.query.message;
        }
    }
    return botEvent;
}

exports.toString = function (obj) {
    if (typeof obj === "undefined") {
        return null;
    } else if (typeof obj === "number") {
        return obj + "";
    } else if (typeof obj === "boolean") {
        return obj + "";
    } else if (typeof obj === "string") {
        return obj;
    } else if (typeof obj === "object") {
        return JSON.stringify(obj, null, 2);
    } else {
        return null;
    }
}

exports.assignAWSCreds = function (AWS) {
    AWS.config.credentials = global.aws_creds;
    return AWS;
}

exports.getAWSSDKInstance = function () {
    return global.aws || require("aws-sdk");
}
